runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/BoundedTrieData.java

```diff
@@ -0,0 +1,583 @@
+   *
+   * @param singleton the singleton list of strings, or null if the data is a trie
+   * @param root the root of the trie, or null if the data is a singleton list
+   * @param bound the maximum number of elements allowed in the trie
+   * @throws IllegalArgumentException if both {@code singleton} and {@code root} are non-null
+   */
+  public BoundedTrieData(
+      @Nullable List<String> singleton, @Nullable BoundedTrieNode root, int bound) {
+    assert singleton == null || root == null;
+    this.singleton = singleton;
+    this.root = root;
+    this.bound = bound;
+  }
+
+  /** Converts this {@link BoundedTrieData} to its proto {@link BoundedTrie}. */
+  public synchronized BoundedTrie toProto() {
+    BoundedTrie.Builder builder = BoundedTrie.newBuilder();
+    builder.setBound(this.bound);
+    if (this.singleton != null) {
+      builder.addAllSingleton(this.singleton);
+    }
+    if (this.root != null) {
+      builder.setRoot(this.root.toProto());
+    }
+    return builder.build();
+  }
+
+  /** Creates a {@link BoundedTrieData} instance from its proto {@link BoundedTrie}. */
+  public static BoundedTrieData fromProto(BoundedTrie proto) {
+    if (proto.hasRoot()) {
+      return new BoundedTrieData(
+          null, BoundedTrieNode.fromProto(proto.getRoot()), proto.getBound());
+    } else {
+      return new BoundedTrieData(proto.getSingletonList(), null, proto.getBound());
+    }
+  }
+
+  /** Returns this {@link BoundedTrieData} as a {@link BoundedTrieNode}. */
+  @Nonnull
+  private synchronized BoundedTrieNode asTrie() {
+    if (this.root != null) {
+      return this.root;
+    } else {
+      BoundedTrieNode trieNode = new BoundedTrieNode();
+      if (this.singleton != null) {
+        trieNode.add(this.singleton);
+      }
+      return trieNode;
+    }
+  }
+
+  /** Returns a new {@link BoundedTrieData} instance that is a deep copy of this instance. */
+  public synchronized BoundedTrieData getCumulative() {
+    List<String> singleton = this.singleton == null ? null : new ArrayList<>(this.singleton);
+    // deep copy
+    BoundedTrieNode root = this.root == null ? null : this.root.deepCopy();
+    return new BoundedTrieData(singleton, root, this.bound);
+  }
+
+  /**
+   * Returns {@link BoundedTrieResult}, which represents all path in the bounded trie. The last
+   * element in each path is a boolean in string representation denoting whether this path was
+   * truncated. i.e. <["a", "b", "false"], ["c", "true"]>
+   *
+   * @return The set of paths.
+   */
+  public synchronized BoundedTrieResult extractResult() {
+    if (this.root == null) {
+      if (this.singleton == null) {
+        return BoundedTrieResult.empty();
+      } else {
+        List<String> list = new ArrayList<>(this.singleton);
+        list.add(String.valueOf(false));
+        return BoundedTrieResult.create(Collections.singleton(list));
+      }
+    } else {
+      return BoundedTrieResult.create(new HashSet<>(this.root.flattened()));
```

Comment:
Nit: it seems more natural to me for BoundedTrieResult to contain a BoundedTrieData (or at least take that as a constructor) rather than do the explosion here.